simdpp/algorithm/transform.h

```diff
@@ -0,0 +1,172 @@
+            const auto size_simd_loop = range.second;
+            
+
+            auto i = 0u;
+
+            //---prologue
+            for (; i < size_prologue_loop; ++i)
+            {
+                *out++ = f(*first++);
+            }
+            //---main simd loop
+            if (detail::is_aligned(out, alignment))
+            {
+                for (; i < size_simd_loop; i += simd_size)
+                {
+                    simd_type_T element = load(first);
+                    store(out, f(element));
+                    first += simd_size;
+                    out += simd_size;
+                }
+            }
+            else
+            {
+                for (; i < size_simd_loop; i += simd_size)
+                {
+                    simd_type_T element = load(first);
+                    store_u(out, f(element));
+                    first += simd_size;
+                    out += simd_size;
+                }
+            }
+            //---epilogue
+            for (; i < size; ++i)
+            {
+                *out++ = f(*first++);
+            }
+            return out;
+        }
+        template<typename T1, typename T2, typename U, typename BinOp>
+        U* transform(T1 const* first1, T1 const* last1, T2 const* first2, U* out, BinOp f)
+        {
+#ifndef NDEBUG //precondition debug mode           
+            if (!first1)
+                throw std::runtime_error("transform - null ptr first1.");
+            if (!last1)
+                throw std::runtime_error("transform - null ptr last1.");
+            if (!first2)
+                throw std::runtime_error("transform - null ptr first2.");
+            if (!out)
+                throw std::runtime_error("transform - null ptr out.");
+#endif
+            using simd_type_T1 = typename typetraits<T1>::simd_type;
+            using simd_type_T2 = typename typetraits<T2>::simd_type;
+            using simd_type_U = typename typetraits<U>::simd_type;
+
+            static_assert (simd_type_T1::base_length == simd_type_T2::base_length
+                , "mismatch base_length between T1 and T2"
+                );
+            static_assert (simd_type_T1::base_length == simd_type_U::base_length
+                , "mismatch base_length between T1 and U"
+                );
+            static_assert (simd_type_T2::base_length == simd_type_U::base_length
+                , "mismatch base_length between T2 and U"
+                );
+
+            auto alignment = typetraits<T1>::alignment;
+
+            //Define loop counter
+            const auto simd_size = simd_type_T1::base_length;
+            const auto size = std::distance(first1, last1);
+            const auto range = helper_input_range(first1, last1);
+            const auto size_prologue_loop = range.first;
+            const auto size_simd_loop = range.second;
+
+
+            auto i = 0u;
+
+            //---prologue
+            for (; i < size_prologue_loop; ++i)
```

Comment:
I think it makes sense to do possibly overlapping SIMD operations for the prologue and epilogue. Some elements will be computed twice, but the code will be much faster this way.

Reply:
idem